### PR TITLE
Schedule add reschecule and cancel

### PR DESCRIPTION
**Pull request by zjb0807**
Closes: #701

### CODE REVIEW

**Review by xlc — COMMENTED** [review]
https://github.com/paritytech/substrate/blob/29aca981db5e8bf8b5538e6c7920ded917013ef3/frame/scheduler/src/lib.rs#L703-L712

Should be using those methods for cancel / reschedule

File: runtime/common/src/precompile/schedule_call.rs [hunk 1]
```diff
@@ -161,29 +171,78 @@ impl<AccountId, AddressMapping, Scheduler, ChargeTransactionPayment, Call, Origi
...
+					ExitError::Other(err_msg.into())
+				})?;
+
+				#[cfg(not(feature = "with-ethereum-compatibility"))]
+				{
+					//// unreserve the transaction fee for gas_limit
+					ChargeTransactionPayment::unreserve_fee(&from_account, fee);
+				}
+
+				Scheduler::cancel_named(task_id).map_err(|_| ExitError::Other("Cancel schedule failed".into()))?;
```
Comment (zjb0807) [inline runtime/common/src/precompile/schedule_call.rs:208]: Here use the scheduler  to do `cancel/reschedule`

Comment (zjb0807) [reply runtime/common/src/precompile/schedule_call.rs:208]: EVMSchedulerManager just saves the caller and checks whether it has permission.

**Comment by zjb0807** [comment]
It is found that `schedule_named` and `cancel_named` are called with root, and the origin cannot be checked.
Redefine `task_id = id + sender + fee`.
So check origin directly through `task_id`.

File: modules/support/src/lib.rs [hunk 2]
```diff
@@ -325,7 +325,8 @@ pub trait EVMStateRentTrait<AccountId, Balance> {
 }
 
 pub trait TransactionPayment<AccountId, Balance, NegativeImbalance> {
-	fn reserve_fee(who: &AccountId, weight: Weight) -> DispatchResult;
+	fn reserve_fee(who: &AccountId, weight: Weight) -> Result<Balance, DispatchError>;
+	fn unreserve_fee(_who: &AccountId, _fee: Balance);
```
Comment (xlc) [inline modules/support/src/lib.rs:329]:
```suggestion
	fn unreserve_fee(who: &AccountId, fee: Balance);
```

File: runtime/common/src/precompile/schedule_call.rs [hunk 3]
```diff
@@ -155,37 +161,97 @@ impl<AccountId, AddressMapping, Scheduler, ChargeTransactionPayment, Call, Origi
...
 			}
 		}
 	}
 }
 
-fn vec_u8_from_tuple(task_address: TaskAddress<BlockNumber>) -> Vec<u8> {
-	let mut be_bytes_0 = [0u8; 32];
-	U256::from(task_address.0).to_big_endian(&mut be_bytes_0[..]);
+fn join_task_id(id: Vec<u8>, sender: H160, fee: Balance) -> Vec<u8> {
+	let mut be_bytes = [0u8; 96];
```
Comment (xlc) [inline runtime/common/src/precompile/schedule_call.rs:240]: you can do `Encode::encode((id, sender, Compact(fee)))`, and `Decode::<(Vec<u8>, H160, Compact<Balance>)>::decode(input)`

Also use compact to save some bytes

File: runtime/common/src/precompile/schedule_call.rs [hunk 4]
```diff
@@ -155,37 +161,97 @@ impl<AccountId, AddressMapping, Scheduler, ChargeTransactionPayment, Call, Origi
...
+
+				// task_id = id + sender + reserve_fee
+				let (_, sender, fee) = split_task_id(&task_id);
+				ensure!(sender == from, ExitError::Other("NoPermission".into()));
+
+				#[cfg(not(feature = "with-ethereum-compatibility"))]
+				{
+					// unreserve the transaction fee for gas_limit
+					let from_account = AddressMapping::get_account_id(&from);
+					ChargeTransactionPayment::unreserve_fee(&from_account, fee.into());
```
Comment (xlc) [inline runtime/common/src/precompile/schedule_call.rs:203]: this can only happen if cancel is successful. i.e. we have verified the task id is valid

**Review by xlc — CHANGES_REQUESTED** [review]

File: runtime/common/src/precompile/schedule_call.rs [hunk 5]
```diff
@@ -13,17 +13,25 @@ use frame_support::{
...
-use codec::Encode;
+use super::input::{Input, InputT, PER_PARAM_BYTES};
+use codec::{Decode, Encode};
 use pallet_scheduler::TaskAddress;
 
 parameter_types! {
 	pub storage EvmSchedulerNextID: u32 = 0u32;
 }
 
+#[derive(Debug, PartialEq, Encode, Decode)]
```
Comment (xlc) [inline runtime/common/src/precompile/schedule_call.rs:27]:
```suggestion
#[derive(RuntimeDebug, PartialEq, Encode, Decode)]
```

File: runtime/common/src/precompile/schedule_call.rs [hunk 6]
```diff
@@ -13,17 +13,25 @@ use frame_support::{
...
+struct TaskInfo {
+	id: Vec<u8>,
```
Comment (xlc) [inline runtime/common/src/precompile/schedule_call.rs:29]:
```suggestion
	prefix: &'static str,
	id: u32,
```

to avoid another encode

**Review by xlc — APPROVED** [review]

File: runtime/common/src/precompile/tests.rs [hunk 7]
```diff
@@ -215,14 +219,65 @@ fn schedule_call_precompile_should_work() {
...
+		let event = TestEvent::pallet_scheduler(pallet_scheduler::RawEvent::Scheduled(3, 0));
+		assert!(System::events().iter().any(|record| record.event == event));
 
-		// returned block + index
-		let mut expected_output = [0u8; 64];
-		U256::from(3).to_big_endian(&mut expected_output[..32]);
-		U256::from(0).to_big_endian(&mut expected_output[32..64]);
+		let mut num = [0u8; 8];
+		num[..].copy_from_slice(&output[32 - 8..32]);
+		let task_id_len: usize = usize::from_be_bytes(num);
```
Comment (xlc) [inline runtime/common/src/precompile/tests.rs:229]: usize is u64 in std and u32 in wasm, so I think should use u32 here

**Pull request merged**